Registry sync packet compression II

deirn opened the pull request:
The compression method is based on #416, with some changes:
- This PR also groups all registry id with same namespace, alongside grouping the object id.
- This PR sends empty string in place of `minecraft` namespace :tiny_potato:
- This PR changes how to send the bulk rawIds. Since most mods registers all at once, the rawId of those objects would be adjacent to each other (the differential would be 1). We can skip sending those differential varints by sending the difference of the first raw id and the last raw id of the bulk before, alongside the size of each bulk. Then let the client loop and increment the raw id.

This PR also keeps the old NBT-based packet to support old server/client.

Number comparison is with [AOF4](https://www.curseforge.com/minecraft/modpacks/all-of-fabric-4/files/3526987) plus [Concistency+](https://www.curseforge.com/minecraft/mc-mods/consistency/files/3510835), which is a mod that has [a fair amount](https://github.com/ConsistencyPlus/ConsistencyPlus/blob/0.4.1/src/main/java/io/github/consistencyplus/consistency_plus/registry/CPlusBlocks.java) of reg objects ;-)
| Packet | Raw size | Deflated size |
| :-: | -: | -: |
| NBT (Old) | `768102` | `210493` |
| Direct (New) | `400934` | `101688` |

Closes #416 
Edit: had a big brain moment and saved some more bytes lol

deirn committed:
Optimize registry sync packet

deirn committed:
Make sure that the id map is sorted by the raw id
Send the difference of the first raw id and the last raw id from the bulk before

liach commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/RegistrySyncManager.java:
```diff
@@ -38,55 +38,70 @@
+			sendPacket(player, DirectRegistrySyncPacket.getInstance());
+		} else {
+			LOGGER.warn("Player {} can't receive direct packet, using nbt packet instead", player.getEntityName());
+			sendPacket(player, NbtRegistrySyncPacket.getInstance());
```
Should we consider kicking the client if it cannot accept the legacy registry sync as well, as otherwise the client will have a glitchy experience?

liach commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/FabricRegistryInit.java:
```diff
@@ -21,10 +21,16 @@
 	@Override
 	public void onInitialize() {
+		ServerPlayConnectionEvents.JOIN.register((handler, sender, server) -> {
+			// TODO: If integrated and local, don't send the packet (it's ignored)
```
You can check if this server is integrated and this sender is server's owner

liach commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/RegistrySyncManager.java:
```diff
@@ -38,55 +38,70 @@
-	static final Identifier ID = new Identifier("fabric", "registry/sync");
 	private static final Logger LOGGER = LogManager.getLogger("FabricRegistrySync");
 	private static final boolean DEBUG_WRITE_REGISTRY_DATA = System.getProperty("fabric.registry.debug.writeContentsAsCsv", "false").equalsIgnoreCase("true");
+	private static final boolean FORCE_NBT_SYNC = System.getProperty("fabric.registry.forceNbtSync", "false").equalsIgnoreCase("true");
```
I think we can just use `Boolean.getBoolean("fabric.registry.forceNbtSync")`

JamiesWhiteShirt commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/packet/RegistrySyncPacket.java:
```diff
@@ -0,0 +1,33 @@
+
+import net.fabricmc.fabric.impl.registry.sync.map.RegistryMap;
+
+public interface RegistrySyncPacket {
```
It is misleading to call this a registry sync packet. Implementations are not packets but rather strategies for sending and serializing registry sync packets.

liach reviewed (commented):
The result of `writeBuffer` should preferably be cached, especially given preparing the direct registry sync buffer is more costly.

JamiesWhiteShirt commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/packet/DirectRegistrySyncPacket.java:
```diff
@@ -0,0 +1,189 @@
+
+				IdMap idMap = map.get(regId);
+
+				// Group object ids with name namespace
+				Map<String, Int2ObjectMap<List<String>>> idNamespaceGroups = new LinkedHashMap<>();
+
+				for (Object2IntMap.Entry<Identifier> idPair : idMap.sortedEntryList()) {
+					Identifier id = idPair.getKey();
+					int rawId = idPair.getIntValue();
+
+					Int2ObjectMap<List<String>> adjacentRawIds = idNamespaceGroups.computeIfAbsent(getNamespace(id), s -> new Int2ObjectLinkedOpenHashMap<>());
+
+					// Group adjacent rawIds together
+					List<String> rawIdGroup;
+
+					if (adjacentRawIds.containsKey(rawId - 1)) {
+						rawIdGroup = adjacentRawIds.remove(rawId - 1);
+					} else {
+						rawIdGroup = new ArrayList<>();
+					}
+
+					rawIdGroup.add(id.getPath());
+					adjacentRawIds.put(rawId, rawIdGroup);
+				}
```
I see a few ways the producing part of the loop that iterates over the registries could be made much more efficient.

Grouping by namespace into a sorted list of entries in the namespace first would allow you to use a more efficient strategy to find consecutive ranges of raw IDs. See [this](https://stackoverflow.com/a/55914488) for some inspiration. It could represent each range as a slice of the sorted list. It would also limit what needs to be in memory to one namespace at a time.

deirn replied:
Changed, how about now?

JamiesWhiteShirt commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/map/IdMap.java:
```diff
@@ -0,0 +1,35 @@
+	/**
+	 * Returns an entry pair list sorted by the rawId.
+	 */
+	public List<Entry<Identifier>> sortedEntryList() {
+		return object2IntEntrySet().stream()
+				.sorted(Comparator.comparingInt(Entry::getIntValue))
+				.toList();
+	}
```
I don't think it is a good idea to extend collection classes. This could just as easily be a method somewhere else that takes the ID map as a parameter.

shartte replied:
agreed

shartte commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/map/RegistryMap.java:
```diff
@@ -0,0 +1,65 @@
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.util.Identifier;
+
+public class RegistryMap extends LinkedHashMap<Identifier, IdMap> {
```
Same as above, I'd not extend collection classes and instead use static helpers

shartte commented:
How are you testing the numbers? I think there might be some other ways to achieve smaller packet sizes by optimizing more for deflate. Can you potentially provide a binary dump of the NBT packet that you used for your tests?

Technici4n commented:
I think we should just bump the max packet size, or deflate the packet ourselves before encoding the packet. I wouldn't spend too much time on something that zlib will do almost as well.

shartte commented:
I'd hypothesize that simply reorganizing the data could already be sufficient in decreasing the compressed size by optimizing for deflate a little. But without being able to run the same extreme case benchmark, I can't test that easily.

deirn committed:
Add packet size log prints

deirn committed:
Use Boolean.getBoolean
Don't send sync packet if the player is the server host

deirn committed:
Use generic Map

deirn commented:
Added log prints to check the buffer size, probable need to revert that commit later.

liach commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/RegistrySyncManager.java:
```diff
@@ -90,6 +97,38 @@ private static void sendPacket(ServerPlayerEntity player, RegistrySyncPacket pac
+		if (DEBUG) {
+			LOGGER.info("{} raw size: {}", packet.getClass().getSimpleName(), buf.readableBytes());
+
+			// see PacketDeflater
```
Hmm, doesn't vanilla deflate custom payloads already? Does deflating again decrease sizes?

deirn replied:
I don't see a way to know what the current packet is being deflated in the server, so instead I manually compress it in the client instead. I think that's what  JamiesWhiteShirt did on their comparison. The client uncompresses the buffer again.

liach replied:
Hmm, if you really want compression, should we build the byte buf and send it into a gzip input stream/output stream for better results, though this might be a silly idea?

deirn replied:
No, this is done to mimic the vanilla compression so we can measure it. That part is copied directly from vanilla code.

deirn committed:
Multiple raw id bulk test

Patbox commented:
> I think we should just bump the max packet size, or deflate the packet ourselves before encoding the packet. I wouldn't spend too much time on something that zlib will do almost as well.

This solution doesn't really fix the problem and will cause issues with vanilla clients (things will error only on vanilla clients, which will make debugging require more work)

Patbox commented:
Looking that you are already refactoring it, it would be nice if you could add some events/etc for excluding entries from syncing. It would be useful for mods wanting to register things server side without requiring mods on client (see [Polymer](https://github.com/Patbox/polymer))

modmuss50 commented:
IMO that’s out of scope for this PR. I have already looked at that here: https://github.com/FabricMC/fabric/pull/1339

deirn committed:
Group object namespace first before grouping raw ids

deirn committed:
Revert "Add packet size log prints"

This reverts commit fcb8b71e

deirn committed:
RegistrySyncPacket -> RegistryPacketSerializer
Move RegistryPacketSerializer singleton to the interface

deirn committed:
Packet size log prints

JamiesWhiteShirt commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/packet/DirectRegistryPacketSerializer.java:
```diff
@@ -0,0 +1,175 @@
+
+				Object2IntMap<Identifier> idMap = map.get(regId);
+
+				// Sort object ids by raw id and group them by namespace
+				Map<String, List<Object2IntMap.Entry<Identifier>>> idNamespaceGroups = idMap.object2IntEntrySet().stream()
+						.sorted(Comparator.comparingInt(Object2IntMap.Entry::getIntValue))
+						.collect(Collectors.groupingBy(e -> e.getKey().getNamespace(), LinkedHashMap::new, Collectors.toList()));
```
It is assumed that each group is sorted by the stream being sorted before being collected into group lists. Consider doing this in the opposite order - group by namespace, then sort each group by raw ID. Sorting after grouping would make the intent clearer. It also reduces the large sort into a few independent smaller sorts, which should be slightly more efficient.

deirn replied:
Upon further reading, `SimpleRegistry`, which is what all registry basically is, is using `rawIdToEntry` list as backing for its iterator, making it already sorted by the raw id. Should we just don't bother re-sorting it here? I already used linked map for the id map in `RegistrySyncManager#createAndPopulateRegistryMap` anyway.

JamiesWhiteShirt replied:
Even if it is the case that SimpleRegistry's iteration order is by ascending raw ID, I don't know if it should be considered part of its Iterable contract or just an implementation detail that may change in the future. We would also have to ensure that items continue to be inserted into the linked map in that order in case we ever want to change `createAndPopulateRegistryMap` or the surrounding logic that provides the parameter to `writeBuffer`. If we are going to rely on this, it should at least be documented.

deirn committed:
Sort raw id on each namespace group

deirn committed:
Use Identifier.DEFAULT_NAMESPACE

oliviathevampire commented:
Progress on this? I'm in huge need for it

modmuss50 reviewed (approved):
Pending a lot of testing, I think this looks good.

I think there is minimal risk to world saves with this as its just syncing, and not the saving part of it.

Technici4n commented:
Shouldn't this PR also tackle splitting the packet if it exceeds a certain size? Otherwise we will encounter the same issue again in the future I think.

modmuss50 commented:
Thats a fair point, its technically not a full solution. Is PR a realistic limit that no one will ever get past? Idk.

At the very least it could accept multiple packets so could have 1 large packet for 1 large registry?

Technici4n commented:
This brings roughly a x2 improvement given the numbers in the description, which is what it is, but not enough to ensure the issue is solved forever.

deirn commented:
From what I can tell, the 2MB limit that causes #410 is only for `readNbt`, so the direct method won't trigger that. 
Not really sure what `0x800000` number is in `PacketInflater` but that looks like the limit for the packet itself (8MB if so).

deirn commented:
Hmm, nevermind, `CustomPayloadS2CPacket` limits the packet to 1MB, making it easier to hit :/

JamiesWhiteShirt commented:
When talking about size limitations, make sure to clarify if you mean compressed size or uncompressed size.

I don't remember the specifics, but there are indeed two different limitations. We've seen registry sync packets containing an NBT payload exceeding 2MB (I believe) uncompressed, which is the limit for `readNbt`. I believe at some point a new overload for `readNbt` appeared which could bypass this limit. It could have worked as a stopgap, but we have a proper solution now.

If there is a limit of 1MB for CustomPayloadS2CPacket, I'm guessing that's 1MB compressed, otherwise we wouldn't be able to hit the `readNbt` limit. The current NBT implementation deflates quite well, so it could easily fit a >2MB uncompressed payload in less than 1MB compressed. Or could the size limit for CustomPayloadS2CPacket have changed since we last observed [this](https://github.com/FabricMC/fabric/issues/410#issuecomment-544383443)?

----

Either way, if we are going to future proof this, I would suggest sending one packet per registry as a natural split.

deirn committed:
Slice to multiple packet if necessary

deirn committed:
Make MAX_PAYLOAD_SIZE configurable
Make testmod registry bigger
Print buffer size in equality test

deirn commented:
Alright, it now split into multiple packets **only** if it's necessary. The testmod registry won't exceed 1MB though to make the equality check between nbt and direct packet works, so I added a configurable max payload size (defaults to 1MB).

The way it split packets is to slice it to multiple `MAX_PAYLOAD_SIZE`d buffer and reconstruct the whole buffer in the client. It also send an empty buffer to mark the end of the split.

---
> If there is a limit of 1MB for CustomPayloadS2CPacket, I'm guessing that's 1MB compressed,

The size check is performed in the constructor, so it's probably the uncompressed buffer. I think Mojang added those in 1.17, idk why.

JamiesWhiteShirt commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/packet/DirectRegistryPacketHandler.java:
```diff
@@ -0,0 +1,244 @@
+ * </ul>
+ */
+public class DirectRegistryPacketHandler extends RegistryPacketHandler {
+	public static final RegistryPacketHandler INSTANCE = new DirectRegistryPacketHandler();
```
Small nitpick: With `DirectRegistryPacketHandler` now being stateful, it should not be singleton-ish. The `DirectRegistryPacketHandler` instance used at runtime should probably be placed elsewhere, such as in a static field in `RegistrySyncManager`.

deirn replied:
Done :+1:

deirn committed:
Move static packet handler instance to RegistrySyncManager
Release combinedBuf before discarding it
Discard syncedRegistryMap after it's accessed

Player3324 reviewed (approved):
LGTM, thanks!

deirn committed:
Use PacketByteBufs to create buffers
